### PR TITLE
VsHierarchy refactoring

### LINKED ISSUE #9953: Performance: Reduce UIDelays in GetPackageSpecsAsync for Legacy projects and non PR projects
There are plenty of UI delays causes in this codepaths that affect codepaths at initialization and more importantly restore.

There are many related issues here, so consider this an epic if necessary. 

The goal here is to reduce the impact of legacy project on all NuGet operations. 

Issues: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1168242

### PR DESCRIPTION
**Pull request by srdjanjovcic**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9953 (partially)
Regression: No

## Fix

Because my previous attempt to asynchronize access to DTE and IVsHierarchy was too much, and failed some tests, lets break it into steps. This part is purely mechanical refactoring to bring related code together into one place and prepare for actual fixes.

## Testing/Validation

Tests Added: No. 
Reason for not adding tests: Just moving code around and minor stylistic changes. Existing tests pass.
Validation: Automated testing.

### CODE REVIEW

**Comment by zivkan** [comment]
I don't mind the changes, but I don't want to set a precedent of having stylistic-only PRs.  I don't want to spend time reviewing many PRs with only style changes, distracting me from customer impacting changes. Worse would be if there are two people with differences in stylistic preference and keep creating PRs to change each others work.

**Comment by srdjanjovcic** [comment]
@zivkan, I agree with you, and I realized that I went a bit much with cleanup 🙂... I amended PR to have some minimal changes to separate concerns, before I dig more into functional changes.

**Review by nkolev92 — APPROVED** [review]

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/EnvDteProjectExtensions.cs [hunk 1]
```diff
@@ -4,23 +4,56 @@
...
     {
+        public static IVsHierarchy ToVsHierarchy(this EnvDTE.Project project)
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            IVsHierarchy hierarchy;
+
+            // Get the vs solution
+            var solution = ServiceLocator.GetInstance<IVsSolution>();
+            var hr = solution.GetProjectOfUniqueName(project.GetUniqueName(), out hierarchy);
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/EnvDteProjectExtensions.cs]: explicit type

File: src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/EnvDteProjectExtensions.cs [hunk 2]
```diff
@@ -4,23 +4,56 @@
...
+
+        public static string[] GetProjectTypeGuids(this EnvDTE.Project project)
+        {
+            Verify.ArgumentIsNotNull(project, nameof(project));
+
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            // Get the vs hierarchy as an IVsAggregatableProject to get the project type guids
+            var hierarchy = ToVsHierarchy(project);
+            var projectTypeGuids = VsHierarchyUtility.GetProjectTypeGuids(hierarchy, project.Kind);
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/IDE/EnvDteProjectExtensions.cs]: explicit types

File: src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/ProjectNames.cs [hunk 3]
```diff
@@ -90,9 +90,9 @@ public static async Task<ProjectNames> FromDTEProjectAsync(EnvDTE.Project dtePro
             await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
             var fullname = dteProject.FullName;
-            var uniqueName = EnvDTEProjectInfoUtility.GetUniqueName(dteProject);
-            var shortName = EnvDTEProjectInfoUtility.GetName(dteProject);
-            var customUniqueName = await EnvDTEProjectInfoUtility.GetCustomUniqueNameAsync(dteProject);
+            var uniqueName = dteProject.GetUniqueName();
+            var shortName = dteProject.GetName();
+            var customUniqueName = await dteProject.GetCustomUniqueNameAsync();
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Common/ProjectSystems/ProjectNames.cs]: explicit types

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs [hunk 4]
```diff
@@ -184,7 +184,7 @@ private async Task<NuGetProject> CreateNuGetProjectAsync(string projectUniqueNam
 
             foreach (Project solutionProject in supportedProjects)
             {
-                var solutionProjectPath = EnvDTEProjectInfoUtility.GetFullProjectPath(solutionProject);
+                var solutionProjectPath = solutionProject.GetFullProjectPath();
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPathContextProvider.cs]: explicit type

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstaller.cs [hunk 5]
```diff
@@ -69,7 +69,7 @@ private void RunJTFWithCorrectContext(Project project, Func<Task> asyncTask)
                 {
                     await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
 
-                    var vsHierarchy = VsHierarchyUtility.ToVsHierarchy(project);
+                    var vsHierarchy = project.ToVsHierarchy();
```
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/Extensibility/VsPackageInstaller.cs]: explicit type (I think?)

File: test/TestExtensions/API.Test/InternalAPITestHook.cs [hunk 6]
```diff
@@ -173,7 +173,7 @@ public static bool IsFileExistsInProject(string projectUniqueName, string filePa
 
                     foreach (EnvDTE.Project project in dte.Solution.Projects)
                     {
-                        var solutionProjectPath = EnvDTEProjectInfoUtility.GetFullProjectPath(project);
+                        var solutionProjectPath = project.GetFullProjectPath();
```
Comment (donnie-msft) [inline test/TestExtensions/API.Test/InternalAPITestHook.cs]: explicit type

**Review by donnie-msft — APPROVED** [review]
Just a few coding guideline-related changes. 
I like the idea of moving to extensions from Utility overall.

**Pull request merged**